Stratis.Bitcoin.Tests/BlockPulling/DownloadAssignmentStrategyTest.cs

```diff
@@ -0,0 +1,106 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Xunit;
+using Stratis.Bitcoin.BlockPulling;
+
+namespace Stratis.Bitcoin.Tests.BlockPulling
+{
+    /// <summary>
+    /// Tests of DownloadAssignmentStrategy class.
+    /// </summary>
+    public class DownloadAssignmentStrategyTest
+    {
+        /// <summary>
+        /// Previous implementation of block puller's strategy could lead to a situation in which the node's 
+        /// peers were asked for blocks then did not have. This is undesirable.
+        /// <para>
+        /// We simulate the following scenario in this test:
+        /// <list type="bullet">
+        /// <item>Our node has a chain with 5 blocks and is connected to 4 peer nodes - A, B, C, D.</item>
+        /// <item>Node A has a chain with 4 blocks.</item>
+        /// <item>Node B has a chain with 20 blocks.</item>
+        /// <item>Node C has a chain with 30 blocks.</item>
+        /// <item>Node D has a chain with 40 blocks.</item>
+        /// </list>
+        /// </para>
+        /// <para>
+        /// We call AskBlocks on the block puller with requests to download blocks 6 to 40
+        /// and we check that the node A is not assigned any work and that the node B is not 
+        /// assigned any work for blocks 21 to 40, and C is not assigned any work for blocks 
+        /// 31 to 40.
+        /// </para>
+        /// </summary>
+        [Fact]
+        public void AssignBlocksToPeersWithNodesWithDifferentChainsCorrectlyDistributesDownloadTasks()
+        {
+            // Create list of numbers 6 to 40 and shuffle it.
+            Random rnd = new Random();
+            List<int> requiredBlockHeights = new List<int>();
+            for (int i = 6; i <= 40; i++)
+                requiredBlockHeights.Add(i);
+
+            requiredBlockHeights = requiredBlockHeights.OrderBy(a => rnd.Next()).ToList();
```

Comment:
The idea was to use randomization to avoid testing a single case, which could possibly be designed poorly and thus could lead to a false positive test. 

On the other hand, one does want to prevent situation in which randomization causes some runs of the test to provide different results than others. This is why I went for (hopefully) large enough sample so that chance of passing the test because of "good" order would be minimal. 

If this seems insufficient, we can certainly increase the number of blocks to virtually any reasonable number (i.e. going 100x current counts would do no harm).